src/CoreWCF.RabbitMQ/src/CoreWCF/Channels/RabbitMqConnectionSettings.cs

```diff
@@ -0,0 +1,177 @@
+            }
+
+            var lastSegment = uri.Segments.LastOrDefault();
+            if (lastSegment.EndsWith("/"))
+            {
+                // Exchange found but no queueName, so create unique queueName
+                return $"{TempQueuePrefix}{new Guid()}";
+            }
+
+            return lastSegment.Replace("/", string.Empty);
+        }
+
+        private static string GetExchangeFromUri(Uri uri)
+        {
+            if (uri.Segments.Length < 2)
+            {
+                throw new ArgumentException($"Expected a non-empty segment in the uri {uri}");
+            }
+
+            var secondSegment = uri.Segments[1];
+            if (secondSegment.EndsWith("/"))
+            {
+                // Exchange name was found
+                return secondSegment.Replace("/", string.Empty);
+            }
+
+            // Exchange name was not found
+            return string.Empty;
+        }
+
+        private static string GetRoutingKeyFromUri(Uri uri, string queueName)
+        {
+            if (string.IsNullOrEmpty(uri.Fragment))
+            {
+                return queueName;
+            }
+
+            return uri.Fragment.Replace("#", string.Empty);
+        }
+
+        private static string SetUserName(Uri uri, ICredentials credentials)
+        {
+            var userNameFromCredentials = credentials?.GetCredential(uri, string.Empty).UserName;
+            if (!string.IsNullOrEmpty(userNameFromCredentials))
+            {
+                return userNameFromCredentials;
+            }
+
+            return GetUserNameFromUri(uri);
+        }
+
+        private static string GetUserNameFromUri(Uri uri)
+        {
+            if (string.IsNullOrEmpty(uri.UserInfo))
+            {
+                return ConnectionFactory.DefaultUser;
+            }
+
+            return uri.UserInfo.Split(':').First();
+        }
+
+        private static string SetPassword(Uri uri, ICredentials credentials, SslOption sslOption)
+        {
+            if (!string.IsNullOrEmpty(sslOption.CertPath) || sslOption.Certs?.Count > 0)
+            {
+                return string.Empty;
+            }
+
+            var passwordFromCredentials = credentials?.GetCredential(uri, string.Empty).Password;
+            if (!string.IsNullOrEmpty(passwordFromCredentials))
+            {
+                return passwordFromCredentials;
+            }
+
+            return GetPasswordFromUri(uri);
+        }
+
+        private static string GetPasswordFromUri(Uri uri)
+        {
+            if (string.IsNullOrEmpty(uri.UserInfo) || uri.UserInfo.Split(':').Length < 2)
+            {
+                return ConnectionFactory.DefaultPass;
+            }
+
+            return uri.UserInfo.Split(':').Last();
+        }
+
+        private static SslOption ConfigureSslOption(SslOption sslOption, Uri uri)
+        {
+            if (sslOption != null)
+            {
```

Comment:
As uri can't be null as it's needed in other methods, what happens if the passed in sslOption doesn't match the uri? Do we care? Is it legitimate to have a different ServerName than uri? Is ServerName used to validate the certificate? If that's the case, then that could be the case. What if uri contains SecureAMQPSScheme but SslOption is false?

Reply:
* The idea here was to provide a default sslOption if one isn't provided, but if one is provided, use that even if it's a misconfiguration. 
* ServerName is supposed to match the SAN or CN of the host in the SSL cert [(source)](https://www.rabbitmq.com/ssl.html#dotnet-connection-params). Based on this, my understanding is that the ServerName should match the host uri. Does this sound correct to you?
* If the uri contains SecureAMQPSScheme and SslOption.Enabled is false, connection to the RabbitMQ host will fail (`BrokerUnreachableException`)